src/proxy.c

```diff
@@ -30,8 +30,18 @@ SEXP vec_proxy(SEXP x) {
 
 // [[ register(); include("vctrs.h") ]]
 SEXP vec_proxy_equal(SEXP x) {
-  return vec_proxy_recursive(x, vctrs_proxy_equal);
+  SEXP proxy = PROTECT(vec_proxy_recursive(x, vctrs_proxy_equal));
+  SEXP res = vec_proxy_unwrap(proxy);
+  UNPROTECT(1);
+  return res;
 }
+SEXP vec_proxy_unwrap(SEXP x) {
+  if (is_data_frame(x) && XLENGTH(x) == 1) {
```

Comment:
Maybe faster would be `if (TYPEOF(x) == VECSXP && XLENGTH(x) == 1 && is_data_frame(X))`?

`is_data_frame()` isn't really expensive, but it would probably take longer than this?

I think we use `vec_proxy_equal()` a decent bit, and its used on atomics, so this should be fast for those

Reply:
Just pushed this change.